### PR TITLE
[api-minor] Fix #7798: Refactor scratch canvas usage.

### LINKED ISSUE #7798: Refactor scratch canvas usage
At display/canvas.js, we create DOM canvas. It will be nice if we will be able to replace any DOM document.createElement('canvas') calls with some type of factory to allow it's replacement e.g. at the node.js apps. See also #7525 discussion.

### PR DESCRIPTION
**Pull request by mukulmishra18**
Fixes #7798, Refactoring scratch canvas usage to use both DOM canvas and npm canvas.

### CODE REVIEW

File: src/display/dom_utils.js [hunk 1]
```diff
@@ -33,6 +33,44 @@ var createValidAbsoluteUrl = sharedUtil.createValidAbsoluteUrl;
 
 var DEFAULT_LINK_REL = 'noopener noreferrer nofollow';
 
+if (typeof require === 'function') {
+  var Canvas = require('canvas');
+}
```
Comment (yurydelendik) [inline src/display/dom_utils.js]: remove this require() call  -- `Canvas` is not used anywhere

File: src/display/webgl.js [hunk 2]
```diff
@@ -83,7 +84,8 @@ var WebGLUtils = (function WebGLUtilsClosure() {
     if (currentGL) {
       return;
     }
-    currentCanvas = document.createElement('canvas');
```
Comment (yurydelendik) [inline src/display/webgl.js]: revert the change here -- it is used in the webgl context, you can add a comment that CanvasRactory is not needed here

File: package.json [hunk 3]
```diff
@@ -2,6 +2,7 @@
   "name": "pdf.js",
   "version": "0.8.0",
   "devDependencies": {
+    "canvas": "^1.6.2",
```
Comment (yurydelendik) [inline package.json]: remove this package - we are not ready to use node.js canvas just yet

File: src/display/text_layer.js [hunk 4]
```diff
@@ -186,7 +187,8 @@ var renderTextLayer = (function renderTextLayerClosure() {
       return;
     }
 
-    var canvas = document.createElement('canvas');
```
Comment (yurydelendik) [inline src/display/text_layer.js]: revert the change here -- it is used to measure text length in the DOM, you can add a comment that CanvasRactory is not needed here

File: src/display/font_loader.js [hunk 5]
```diff
@@ -216,7 +218,8 @@ if (typeof PDFJSDev === 'undefined' || !PDFJSDev.test('MOZCENTRAL')) {
 
       var i, ii;
 
-      var canvas = document.createElement('canvas');
```
Comment (yurydelendik) [inline src/display/font_loader.js]: revert the change here -- it is used for hack to determine if font is loaded, you can add a comment that CanvasRactory is not needed here

File: src/display/canvas.js [hunk 6]
```diff
@@ -1454,7 +1449,8 @@ var CanvasGraphics = (function CanvasGraphicsClosure() {
     get isFontSubpixelAAEnabled() {
       // Checks if anti-aliasing is enabled when scaled text is painted.
       // On Windows GDI scaled fonts looks bad.
-      var ctx = document.createElement('canvas').getContext('2d');
+      var canvasFactory = new CanvasFactory();
```
Comment (yurydelendik) [inline src/display/canvas.js]: CanvasFactory must be created once per render() call and stored at CanvasGraphics.

**Review by yurydelendik — CHANGES_REQUESTED** [review]

File: src/display/dom_utils.js (hunk 1 above)
Comment (mukulmishra18) [inline src/display/dom_utils.js]: Actually it is used in the else block of create method, line 48.

File: src/display/canvas.js [hunk 7]
```diff
@@ -478,6 +472,7 @@ var CanvasGraphics = (function CanvasGraphicsClosure() {
     this.smaskCounter = 0;
     this.tempSMask = null;
     this.cachedCanvases = new CachedCanvases();
+    this.DOMcanvasFactory = new DOMcanvasFactory();
```
Comment (Snuffleupagus) [inline src/display/canvas.js]: It may be better to call this `this.canvasFactory`, since I believe that one of the points of issue #7798 is to allow an API consumer to provide their own canvas factory (e.g. using Node canvas), and then a using the `this.DOMCanvasFactory` name here could be confusing.

File: src/display/canvas.js [hunk 8]
```diff
@@ -218,7 +212,7 @@ var CachedCanvases = (function CachedCanvasesClosure() {
         // reset canvas transform for emulated mozCurrentTransform, if needed
         canvasEntry.context.setTransform(1, 0, 0, 1, 0, 0);
       } else {
-        var canvas = createScratchCanvas(width, height);
+        var canvas = CanvasGraphics.DOMcanvasFactory.create(width, height);
```
Comment (Snuffleupagus) [inline src/display/canvas.js]: In order to prevent errors, such as `TypeError: CanvasGraphics.DOMcanvasFactory is undefined[Learn More]  canvas.js:215:13`, you'll need to pass in a `canvasFactory` instance when initializing `CachedCanvases`, i.e. change the "constructor" to 
```javascript
function CachedCanvases(canvasFactory) {
     this.canvasFactory = this.canvasFactory;
     this.cache = Object.create(null);
   }
}
```
and then use `this.canvasFactory.create(width, height)` here.

File: src/display/canvas.js (hunk 7 above)
Comment (Snuffleupagus) [inline src/display/canvas.js]: Furthermore, I think that you also need to initialize the `DOMCanvasFactory` *before* `CachedCanvases`, since you'll probably need to do `this.cachedCanvases = new CachedCanvases(this.canvasFactory);`.
Please refer to https://github.com/mozilla/pdf.js/pull/8002/files#r98327854.

File: src/display/dom_utils.js [hunk 9]
```diff
@@ -33,6 +33,36 @@ var createValidAbsoluteUrl = sharedUtil.createValidAbsoluteUrl;
 
 var DEFAULT_LINK_REL = 'noopener noreferrer nofollow';
 
+function DOMCanvasFactory() {}
```
Comment (Snuffleupagus) [inline src/display/dom_utils.js]: I'm not sure if using `this` the way that you do here below actually makes sense in the context of a factory!? I'd think that you may need to do something like this instead:
```javascript
function DOMCanvasFactory() {}
DOMCanvasFactory.prototype = {
  create: function DOMCanvasFactory_create(width, height) {
    var canvas = document.createElement('canvas');
    canvas.width = width || 0;
    canvas.height = height || 0;

    return canvas;
  },

  reset: function DOMCanvasFactory_reset(canvas, width, height) {
    if (!canvas) {
      throw new Error('Canvas not found');
    }
    canvas.width = width;
    canvas.height = height;

    return canvas;
  },

  destroy: function DOMCanvasFactory_destroy(canvas) {
    if (!canvas) {
      throw new Error('Canvas not found');
    }
    // Zeroing the width and height causes Firefox to release graphics
    // resources immediately, which can greatly reduce memory consumption.
    canvas.width = 0;
    canvas.height = 0;

    canvas = null;
  }
};
```

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
I've added a couple of review comments based on cursory looking at/testing the code.
Having briefly tested this locally, using the default `tracemonkey.pdf` file with `gulp server`, it seems that the current version of this patch unfortunately doesn't work as is.

Again, please remember to run tests (and use the viewer locally), since that helps catch errors as described below; see e.g. https://github.com/mozilla/pdf.js/wiki/Contributing#4-run-lint-and-testing.

File: src/display/canvas.js [hunk 10]
```diff
@@ -50,6 +50,7 @@ var TilingPattern = displayPatternHelper.TilingPattern;
 var getShadingPatternFromIR = displayPatternHelper.getShadingPatternFromIR;
 var WebGLUtils = displayWebGL.WebGLUtils;
 var hasCanvasTypedArrays = displayDOMUtils.hasCanvasTypedArrays;
+var DOMcanvasFactory = displayDOMUtils.DOMCanvasFactory;
```
Comment (mukulmishra18) [inline src/display/canvas.js]: Okay, thanks.

File: src/display/canvas.js (hunk 8 above)
Comment (mukulmishra18) [inline src/display/canvas.js]: I only ran gulp unittest and lint, but forgot to test the viewer. Thanks for pointing out, i will take care of that.

Comment (Snuffleupagus) [inline src/display/canvas.js]: Please don't forget to also run the *entire* test-suite, using `gulp test`, to ensure that there's no rendering regressions.

Comment (mukulmishra18) [inline src/display/canvas.js]: Okay, thanks.

File: src/display/dom_utils.js (hunk 9 above)
Comment (mukulmishra18) [inline src/display/dom_utils.js]: Sorry, i think i was little messed up with constructor function pattern. I will change it.

File: src/display/dom_utils.js [hunk 11]
```diff
@@ -33,6 +33,36 @@ var createValidAbsoluteUrl = sharedUtil.createValidAbsoluteUrl;
...
+    if (this.destroyed) {
+      throw new Error('Canvas not found');
+    }
+    this.canvas.width = width;
+    this.canvas.height = height;
+
+    return this.canvas;
+  },
+
+  destroy: function DOMCanvasFactory_destroy() {
```
Comment (mukulmishra18) [inline src/display/dom_utils.js]: > Please also remember to set the width/height of the canvas to 0 before removing it, since that will cause (at least) Firefox to free graphics resources quicker

That is a good point, i didn't think about that. Thanks for pointing.

File: src/display/canvas.js (hunk 7 above)
Comment (mukulmishra18) [inline src/display/canvas.js]: > It may be better to call it this.canvasFactory, since I believe that one of the points of issue #7798 is to allow an API consumer to provide their own canvas factory (e.g. using Node canvas), and then a using the this.DOMCanvasFactory name here would be confusing.

Should i have to change `this.DOMcanvasFactory` with `this.canvasFactory` in the whole canvas.js file or just the specified place?

Comment (Snuffleupagus) [inline src/display/canvas.js]: I'm not sure that I really understand what you're asking here, but you should (obviously) be consistent within the file. I.e. the above ought to look like below (see also the code snippet in https://github.com/mozilla/pdf.js/pull/8002#discussion_r98327854).
```javascript
this.canvasFactory = new DOMCanvasFactory();
this.cachedCanvases = new CachedCanvases(this.canvasFactory);
```

**Review by Snuffleupagus — DISMISSED** [review]
This is looking much better now!
_**Edit:** It seems that this patch causes **a lot** of movement/failures in some of the tests, and I think I understand why. Please see the inline comment below._

However, I still think that there's one crucial thing left to do here. The way that I interpret issue #7798 is that we need to make it possible for users of the API to provide their own *custom* `CanvasFactory`, for example one using Node.js, for this refactoring to be truly useful.
~~I obviously don't know what @yurydelendik had in mind, but I'd think you probably want something along the lines of the following diff (created on top of this patch):~~ _**Edit:** Based on comments below, this isn't exactly what we want here; hence I'll remove the diff to not cause unnecessary confusion._

File: src/display/dom_utils.js [hunk 12]
```diff
@@ -33,6 +33,37 @@ var createValidAbsoluteUrl = sharedUtil.createValidAbsoluteUrl;
 
 var DEFAULT_LINK_REL = 'noopener noreferrer nofollow';
 
+function DOMCanvasFactory() {}
+DOMCanvasFactory.prototype = {
+  create: function DOMCanvasFactory_create(width, height) {
+    var canvas = document.createElement('canvas');
+    canvas.width = width || 0;
+    canvas.height = height || 0;
```
Comment (Snuffleupagus) [inline src/display/dom_utils.js]: Letting the `width`/`height` breaks `get isFontSubpixelAAEnabled()` case, with *lots* of test failures as a result. I think that you need to do e.g. this instead:
```javascript
if (typeof width !== 'undefined') {
  canvas.width = width;
}
if (typeof height !== 'undefined') {
  canvas.height = height;
}
```

File: src/display/api.js [hunk 13]
```diff
@@ -1742,7 +1742,8 @@ var WorkerTransport = (function WorkerTransportClosure() {
             var size = width * height;
             var rgbaLength = size * 4;
             var buf = new Uint8Array(size * components);
-            var tmpCanvas = createScratchCanvas(width, height);
+            var canvasFactory = new DOMCanvasFactory();
```
Comment (mukulmishra18) [inline src/display/api.js]: [Here](http://logs.glob.uno/?c=mozilla%23pdfjs&s=27+Jan+2017&e=27+Jan+2017#c56478) is the conversation with @yurydelendik, when i am trying to solve this issue. I think i got confused of passing `canvasFactory` via `CanvasGraphics`. I will fix it. Thanks for clearing.

File: src/display/font_loader.js [hunk 14]
```diff
@@ -216,6 +216,8 @@ if (typeof PDFJSDev === 'undefined' || !PDFJSDev.test('MOZCENTRAL')) {
 
       var i, ii;
 
+      // CanvasFactory is not needed here, as it is used to determine if fonts
+      // is loaded.
```
Comment (mukulmishra18) [inline src/display/font_loader.js]: Oh yes, thanks for catch. I will fix it.

File: src/display/dom_utils.js (hunk 12 above)
Comment (mukulmishra18) [inline src/display/dom_utils.js]: Okay, Thanks. I will do it this way.

Comment (yurydelendik) [inline src/display/dom_utils.js]: actually, the code at isFontSubpixelAAEnabled was just creating canvas with default size, we just need to remove `|| 0` or `typeof width !== 'undefined'` and specify needed canvas size at the caller (isFontSubpixelAAEnabled) e.g. 200x100 or 50x10

File: src/display/api.js (hunk 13 above)
Comment (yurydelendik) [inline src/display/api.js]: I think we need to keep this part without change as well and add `assert(typeof document !== 'undefined')`.

I'm thinking the canvasFactory must be a property of render parameters. To protect core from calling browser decoding (which will fail anyway at setting image src), we need to pass a flag to evaluator that shows that browser decoding is disabled. I thought we do/did it, since pdf2svg example worked, but I cannot find this code now, so we need to pass this flag with getOperatorList requests.

File: src/display/canvas.js [hunk 15]
```diff
@@ -1454,7 +1450,7 @@ var CanvasGraphics = (function CanvasGraphicsClosure() {
     get isFontSubpixelAAEnabled() {
       // Checks if anti-aliasing is enabled when scaled text is painted.
       // On Windows GDI scaled fonts looks bad.
-      var ctx = document.createElement('canvas').getContext('2d');
+      var ctx = this.canvasFactory.create().getContext('2d');
```
Comment (yurydelendik) [inline src/display/canvas.js]: create canvas is size (10, 10) here

File: src/display/dom_utils.js [hunk 16]
```diff
@@ -33,6 +33,37 @@ var createValidAbsoluteUrl = sharedUtil.createValidAbsoluteUrl;
...
+
+    return canvas;
+  },
+
+  reset: function DOMCanvasFactory_reset(canvas, width, height) {
+    if (!canvas) {
```
Comment (yurydelendik) [inline src/display/dom_utils.js]: Use `assert(canvas, 'canvas is not specifed')` here (and below)

add asserts (here and above) for the width and height parameters: `assert(width > 0 && height > 0, 'invalid canvas size')`

File: src/display/api.js (hunk 13 above)
Comment (yurydelendik) [inline src/display/api.js]: It is still will be passed to the CanvasGraphics, but it will be specified at the RenderParameters. And if it is not specified, created at the render().

File: src/display/dom_utils.js (hunk 12 above)
Comment (mukulmishra18) [inline src/display/dom_utils.js]: Then i think we should remove `typeof width !== 'undefined'`, as `|| 0` is little bit more clean.

File: src/display/canvas.js (hunk 15 above)
Comment (mukulmishra18) [inline src/display/canvas.js]: got it, thanks

File: src/display/dom_utils.js (hunk 12 above)
Comment (yurydelendik) [inline src/display/dom_utils.js]: yes, meant to say need to remove `|| 0` *and* `typeof width !== 'undefined'`

File: src/display/api.js [hunk 17]
```diff
@@ -1742,7 +1751,10 @@ var WorkerTransport = (function WorkerTransportClosure() {
             var size = width * height;
             var rgbaLength = size * 4;
             var buf = new Uint8Array(size * components);
-            var tmpCanvas = createScratchCanvas(width, height);
+            // make sure that this code is not executing in node.js.
```
Comment (timvandermeij) [inline src/display/api.js]: Make this comment a bit more specific, i.e., why is it bad if this runs with Node.js?

File: src/display/api.js [hunk 18]
```diff
@@ -1742,7 +1751,10 @@ var WorkerTransport = (function WorkerTransportClosure() {
...
+            assert(typeof document !== 'undefined');
+            var canvasFactory = new DOMCanvasFactory();
```
Comment (timvandermeij) [inline src/display/api.js]: Can't we use `this.canvasFactory` (as set above) here? If so, we can remove this line and use `var tmpCanvas = this.canvasFactory.create(width, height);` in the line below.

**Review by timvandermeij — COMMENTED** [review]
This patch looks good to me after addressing these comments, but I'll leave the final review for others that know this particular code a bit better than me.

File: src/display/canvas.js [hunk 19]
```diff
@@ -477,7 +473,8 @@ var CanvasGraphics = (function CanvasGraphicsClosure() {
     this.smaskStack = [];
     this.smaskCounter = 0;
     this.tempSMask = null;
-    this.cachedCanvases = new CachedCanvases();
+    this.canvasFactory = canvasFactory || new DOMCanvasFactory();
```
Comment (yurydelendik) [inline src/display/canvas.js]: `canvasFactory` parameter must be always specified (starting at render() method). can you also move `canvasFactory` before `imageLayer` in the parameters list.

File: src/display/dom_utils.js [hunk 20]
```diff
@@ -26,13 +26,43 @@
...
+  },
+
+  destroy: function DOMCanvasFactory_destroy(canvas) {
+    assert(canvas, 'canvas is not specified');
+    // Zeroing the width and height cause Firefox to release graphics
+    // resources immediately, which can greatly reduce memory consumption.
+    canvas.width = 0;
+    canvas.height = 0;
+
+    canvas = null;
```
Comment (yurydelendik) [inline src/display/dom_utils.js]: remove `canvas = null;`

Also, this method must be used in the CachedCanvases.clear

File: src/display/dom_utils.js [hunk 21]
```diff
@@ -26,13 +26,43 @@
...
 var deprecated = sharedUtil.deprecated;
 var createValidAbsoluteUrl = sharedUtil.createValidAbsoluteUrl;
 
 var DEFAULT_LINK_REL = 'noopener noreferrer nofollow';
 
+function DOMCanvasFactory() {}
+DOMCanvasFactory.prototype = {
+  create: function DOMCanvasFactory_create(width, height) {
+    var canvas = document.createElement('canvas');
+    assert(width > 0 && height > 0, 'invalid canvas size');
```
Comment (yurydelendik) [inline src/display/dom_utils.js]: move assert as a first line of the function

File: src/display/api.js [hunk 22]
```diff
@@ -132,6 +133,8 @@ if (typeof PDFJSDev !== 'undefined' &&
  * @property {string} docBaseUrl - (optional) The base URL of the document,
  *   used when attempting to recover valid absolute URLs for annotations, and
  *   outline items, that (incorrectly) only specify relative URLs.
+ * @property {Object} canvasFactory - (optional) The factory that will be used
```
Comment (yurydelendik) [inline src/display/api.js]: Move `canvasFactory` from here and add to the RenderParameters below. We don't need to store canvas factory on the transport but rather than on a rendering task.

File: src/display/api.js (hunk 18 above)
Comment (yurydelendik) [inline src/display/api.js]: We probable need to make it `var tmpCanvas = document.createElement('canvas'); tmpCanvas.width =...`

**Review by yurydelendik — CHANGES_REQUESTED** [review]

File: src/display/api.js (hunk 17 above)
Comment (mukulmishra18) [inline src/display/api.js]: Okay, thanks. I will do it.

File: src/display/api.js [hunk 23]
```diff
@@ -245,6 +246,7 @@ function getDocument(src, pdfDataRangeTransport,
   }
 
   params.rangeChunkSize = params.rangeChunkSize || DEFAULT_RANGE_CHUNK_SIZE;
+  var canvasFactory = params.canvasFactory || new DOMCanvasFactory();
```
Comment (Snuffleupagus) [inline src/display/api.js]: Based on https://github.com/mozilla/pdf.js/pull/8002#discussion_r99022111, this line should be removed.

File: src/display/api.js [hunk 24]
```diff
@@ -262,7 +264,8 @@ function getDocument(src, pdfDataRangeTransport,
         throw new Error('Loading aborted');
       }
       var messageHandler = new MessageHandler(docId, workerId, worker.port);
-      var transport = new WorkerTransport(messageHandler, task, rangeTransport);
+      var transport = new WorkerTransport(messageHandler, task, rangeTransport,
+                                          canvasFactory);
```
Comment (Snuffleupagus) [inline src/display/api.js]: Based on https://github.com/mozilla/pdf.js/pull/8002#discussion_r99022111, this parameter should be removed.

File: src/display/api.js [hunk 25]
```diff
@@ -720,6 +726,7 @@ var PDFPageProxy = (function PDFPageProxyClosure() {
     this.transport = transport;
     this.stats = new StatTimer();
     this.stats.enabled = getDefaultSetting('enableStats');
+    this.canvasFactory = transport.canvasFactory;
```
Comment (Snuffleupagus) [inline src/display/api.js]: Based on https://github.com/mozilla/pdf.js/pull/8002#discussion_r99022111. this line should be removed.

File: src/display/api.js [hunk 26]
```diff
@@ -1410,10 +1418,12 @@ var PDFWorker = (function PDFWorkerClosure() {
  * @ignore
  */
 var WorkerTransport = (function WorkerTransportClosure() {
-  function WorkerTransport(messageHandler, loadingTask, pdfDataRangeTransport) {
+  function WorkerTransport(messageHandler, loadingTask, pdfDataRangeTransport,
+                           canvasFactory) {
```
Comment (Snuffleupagus) [inline src/display/api.js]: Based on https://github.com/mozilla/pdf.js/pull/8002#discussion_r99022111, this parameter should be removed.

File: src/display/api.js [hunk 27]
```diff
@@ -1410,10 +1418,12 @@ var PDFWorker = (function PDFWorkerClosure() {
...
     this.messageHandler = messageHandler;
     this.loadingTask = loadingTask;
     this.pdfDataRangeTransport = pdfDataRangeTransport;
+    this.canvasFactory = canvasFactory;
```
Comment (Snuffleupagus) [inline src/display/api.js]: Based on https://github.com/mozilla/pdf.js/pull/8002#discussion_r99022111, this line should be removed.

File: src/display/api.js [hunk 28]
```diff
@@ -834,7 +841,8 @@ var PDFPageProxy = (function PDFPageProxyClosure() {
                                                       this.objs,
                                                       this.commonObjs,
                                                       intentState.operatorList,
-                                                      this.pageNumber);
+                                                      this.pageNumber,
+                                                      this.canvasFactory);
```
Comment (Snuffleupagus) [inline src/display/api.js]: Based on https://github.com/mozilla/pdf.js/pull/8002#discussion_r99022111, this line should be replaced by `canvasFactory`, with
```javascript
var canvasFactory = params.canvasFactory || new DOMCanvasFactory();

```
defined at line https://github.com/mozilla/pdf.js/pull/8002/files#diff-0ecad279ed2252e3eb47a4d96ec1463cR815 above.

File: src/display/canvas.js [hunk 29]
```diff
@@ -466,6 +459,7 @@ var CanvasGraphics = (function CanvasGraphicsClosure() {
     this.xobjs = null;
     this.commonObjs = commonObjs;
     this.objs = objs;
+    this.canvasFactory = canvasFactory || new DOMCanvasFactory();
```
Comment (Snuffleupagus) [inline src/display/canvas.js]: Based on https://github.com/mozilla/pdf.js/pull/8002#discussion_r99020974, this line should simply read
```javascript
this.canvasFactory = canvasFactory;
```

File: src/display/canvas.js [hunk 30]
```diff
@@ -456,7 +448,8 @@ var CanvasGraphics = (function CanvasGraphicsClosure() {
   // Defines the number of steps before checking the execution time
   var EXECUTION_STEPS = 10;
 
-  function CanvasGraphics(canvasCtx, commonObjs, objs, imageLayer) {
+  function CanvasGraphics(canvasCtx, commonObjs, objs, imageLayer,
+                          canvasFactory) {
```
Comment (Snuffleupagus) [inline src/display/canvas.js]: Based on https://github.com/mozilla/pdf.js/pull/8002#discussion_r99020974, please change the order of `imageLayer` and `canvasFactory` here.

File: src/display/api.js [hunk 31]
```diff
@@ -2068,7 +2084,8 @@ var InternalRenderTask = (function InternalRenderTaskClosure() {
 
       var params = this.params;
       this.gfx = new CanvasGraphics(params.canvasContext, this.commonObjs,
-                                    this.objs, params.imageLayer);
+                                    this.objs, params.imageLayer,
+                                    this.canvasFactory);
```
Comment (Snuffleupagus) [inline src/display/api.js]: Based on https://github.com/mozilla/pdf.js/pull/8002#discussion_r99020974, please change the order of `params.imageLayer` and `this.canvasFactory` here.

File: src/display/api.js [hunk 32]
```diff
@@ -1742,7 +1752,12 @@ var WorkerTransport = (function WorkerTransportClosure() {
             var size = width * height;
             var rgbaLength = size * 4;
             var buf = new Uint8Array(size * components);
-            var tmpCanvas = createScratchCanvas(width, height);
+            // make sure that this code is not executing in node.js, as
+            // it's using DOM image, and there is no library to support that.
+            assert(typeof document !== 'undefined');
```
Comment (Snuffleupagus) [inline src/display/api.js]: Looking at the entire `JpegDecode` code here, please see https://github.com/mozilla/pdf.js/blob/master/src/display/api.js#L1724-L1769, the use of `assert` seems slightly out of place here.

@yurydelendik I'd think it'd be more in line with the rest of this code if we did
```javascript
if (typeof document === 'undefined') {
  // Make sure that this code is not executing in Node.js, as
  // it's using DOM image, and there is no library to support that.
  return Promise.reject(new Error('"document" is not defined.'));
}
```
at e.g. line https://github.com/mozilla/pdf.js/blob/master/src/display/api.js#L1728 instead.

**Review by Snuffleupagus — COMMENTED** [review]
Sorry about bad advice previously!
Looking through the latest patch, it seems that a couple of things should probably be reverted/changed based on https://github.com/mozilla/pdf.js/pull/8002#discussion_r99022111 and https://github.com/mozilla/pdf.js/pull/8002#discussion_r99020974, so I've added inline comments where applicable.

File: src/display/api.js (hunk 32 above)
Comment (yurydelendik) [inline src/display/api.js]: Either way works for me - it's just to make it easier to catch a problem. We can even move assert/reject at the top of the function.

File: src/display/api.js [hunk 33]
```diff
@@ -61,9 +61,10 @@ var warn = sharedUtil.warn;
 var FontFaceObject = displayFontLoader.FontFaceObject;
 var FontLoader = displayFontLoader.FontLoader;
 var CanvasGraphics = displayCanvas.CanvasGraphics;
-var createScratchCanvas = displayCanvas.createScratchCanvas;
 var Metadata = displayMetadata.Metadata;
 var getDefaultSetting = displayDOMUtils.getDefaultSetting;
+var DOMCanvasFactory = displayDOMUtils.DOMCanvasFactory;
+var assert = sharedUtil.assert;
```
Comment (Snuffleupagus) [inline src/display/api.js]: Remove this line, since `assert` is now unused (which is also why linting failed).

File: src/display/api.js [hunk 34]
```diff
@@ -1727,6 +1733,12 @@ var WorkerTransport = (function WorkerTransportClosure() {
           return Promise.reject(new Error('Worker was destroyed'));
         }
 
+        if (typeof document === 'undefined') {
+          // make sure that this code is not executing in node.js, as
```
Comment (Snuffleupagus) [inline src/display/api.js]: Nit: Please remember to use uppercase letters at the beginning of sentences in comments, i.e. `// Make sure ...` instead.

File: src/display/dom_utils.js [hunk 35]
```diff
@@ -26,13 +26,41 @@
...
 var DEFAULT_LINK_REL = 'noopener noreferrer nofollow';
 
+function DOMCanvasFactory() {}
+DOMCanvasFactory.prototype = {
+  create: function DOMCanvasFactory_create(width, height) {
+    assert(width > 0 && height > 0, 'invalid canvas size');
+    var canvas = document.createElement('canvas');
+    canvas.width = width;
+    canvas.height = height;
+
```
Comment (yurydelendik) [inline src/display/dom_utils.js]: nit: we can remove a empty line here.

File: src/display/dom_utils.js [hunk 36]
```diff
@@ -26,13 +26,41 @@
...
+    return canvas;
+  },
+
+  reset: function DOMCanvasFactory_reset(canvas, width, height) {
```
Comment (yurydelendik) [inline src/display/dom_utils.js]: This function must be used from CachedCanvases

File: src/display/font_loader.js [hunk 37]
```diff
@@ -216,6 +216,8 @@ if (typeof PDFJSDev === 'undefined' || !PDFJSDev.test('MOZCENTRAL')) {
 
       var i, ii;
 
+      // CanvasFactory is not needed here, as it is used to determine if fonts
```
Comment (yurydelendik) [inline src/display/font_loader.js]: Here and below, replace "CanvasFractory is not needed here, as it is" to something like "The temporary canvas is"

**Review by yurydelendik — DISMISSED** [review]
Looks good to me with comments below addressed and tests are passing.

File: src/display/canvas.js [hunk 38]
```diff
@@ -213,12 +207,11 @@ var CachedCanvases = (function CachedCanvasesClosure() {
       var canvasEntry;
       if (this.cache[id] !== undefined) {
         canvasEntry = this.cache[id];
-        canvasEntry.canvas.width = width;
-        canvasEntry.canvas.height = height;
+        this.canvasFactory.reset(canvasEntry, width, height);
```
Comment (Snuffleupagus) [inline src/display/canvas.js]: Are you sure that this actually works, since I believe that you need to use `canvasEntry.canvas` instead (compare with the the `destroy(...)` call below).

Comment (mukulmishra18) [inline src/display/canvas.js]: Oh, okay. I missed something. Sorry

**Comment by yurydelendik** [comment]
Thank you for the patch.

It's very good step, but I think to make it work we need to change a factory to return {canvas, context} pair instead to make it work for node's canvas. @mukulmishra18, would you like to follow up?

**Comment by mukulmishra18** [comment]
@yurydelendik, @Snuffleupagus, @timvandermeij thanks for your reviews and help. You all are very supportive and cooperative, i learned a lot from all of you. I definitely wants to continue, please guide me what i have to do.

**Comment by yurydelendik** [comment]
>  I definitely wants to continue, please guide me what i have to do.

Looking the the way node-canvas is create, we will need to create a new canvas and content on reset/resize. Let's just return create CanvasAndContextPair "structure" from create method and accept it in reset and destroy calls. CachedCanvases will/do store this structure. Notice that every time we create canvas we always create 2d-context after that, so we may do that in the factory.
